lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/runtime/LogicalTimer.java

```diff
@@ -18,36 +18,101 @@
 
+  /**
+   * Instances of the timer should only be created automatically by Tang.
+   */
   @Inject
-  LogicalTimer() {
+  private LogicalTimer() {
   }
 
+  /**
+   * Get current time in milliseconds since the beginning of the epoch (01/01/1970).
+   * This timer implementation always returns the timestamp of the most distant
+   * future event ever checked against this timer in getDuration() or isReady() methods.
+   * Return 0 if there were no calls yet to getDuration() or isReady().
+   * @return Timestamp of the latest event (in milliseconds since the start of the epoch).
+   */
   @Override
   public long getCurrent() {
-    return this.current;
+    return this.current.get();
   }
 
+  /**
+   * Get the number of milliseconds between current time as tracked by the Timer implementation
+   * and a given event. This implementation always returns 0 and updates current timer's time
+   * to the timestamp of the most distant future event.
+   * @param time Timestamp in milliseconds.
+   * @return Always returns 0.
+   * @deprecated [REEF-1532] Prefer passing Time object instead of the numeric timestamp.
+   * Remove after release 0.16.
+   */
   @Override
   public long getDuration(final long time) {
-    isReady(time);
+    this.isReady(time);
     return 0;
   }
 
+  /**
+   * Get the number of milliseconds between current time as tracked by the Timer implementation
+   * and a given event. This implementation always returns 0 and updates current timer's time
+   * to the timestamp of the most distant future event.
+   * @param time Timestamp object that wraps time in milliseconds.
+   * @return Always returns 0.
+   */
+  @Override
+  public long getDuration(final Time time) {
+    return this.getDuration(time.getTimestamp());
+  }
+
+  /**
+   * Check if the event with a given timestamp has occurred, according to the timer.
+   * This implementation always returns true and updates current timer's time to the timestamp
+   * of the most distant future event.
+   * @param time Timestamp in milliseconds.
+   * @return Always returns true.
+   * @deprecated [REEF-1532] Prefer passing Time object instead of the numeric timestamp.
+   * Remove after release 0.16.
+   */
   @Override
   public boolean isReady(final long time) {
-    if (this.current < time) {
-      this.current = time;
+    while (true) {
+      final long thisTs = this.current.get();
+      if (thisTs >= time || this.current.compareAndSet(thisTs, time)) {
+        return true;
+      }
     }
-    return true;
   }
 
+  /**
+   * Check if the event with a given timestamp has occurred, according to the timer.
+   * This implementation always returns true and updates current timer's time to the timestamp
+   * of the most distant future event.
+   * @param time Timestamp object that wraps time in milliseconds.
+   * @return Always returns true.
+   */
+  @Override
+  public boolean isReady(final Time time) {
+    return this.isReady(time.getTimestamp());
```

Comment:
It will be better to validate that the returned value is true, or ignore the returned value and return 'true'